libs/server-sdk/src/data_store/data_store_updater.cpp

```diff
@@ -0,0 +1,78 @@
+#include "data_store_updater.hpp"
+
+#include <launchdarkly/signals/boost_signal_connection.hpp>
+#include <utility>
+
+namespace launchdarkly::server_side::data_store {
+
+std::unique_ptr<IConnection> DataStoreUpdater::OnFlagChange(
+    launchdarkly::server_side::IChangeNotifier::ChangeHandler handler) {
+    std::lock_guard lock{signal_mutex_};
+
+    return std::make_unique<launchdarkly::internal::signals::SignalConnection>(
+        signals_.connect(handler));
+}
+
+void DataStoreUpdater::Init(launchdarkly::data_model::SDKDataSet data_set) {
+    // Optional outside the HasListeners() scope, this allows for the changes
+    // to be calculated before the update and then the notification to be
+    // sent after the update completes.
+    std::optional<DependencySet> change_notifications;
+    if (HasListeners()) {
+        auto updated_items = DependencySet();
+
+        CalculateChanges(DataKind::kFlag, store_->AllFlags(), data_set.flags,
+                         updated_items);
+        CalculateChanges(DataKind::kSegment, store_->AllSegments(),
+                         data_set.segments, updated_items);
+        change_notifications = updated_items;
+    }
+
+    dependency_tracker_.Clear();
+    for (auto const& flag : data_set.flags) {
+        dependency_tracker_.UpdateDependencies(flag.first, flag.second);
+    }
+    for (auto const& segment : data_set.segments) {
+        dependency_tracker_.UpdateDependencies(segment.first, segment.second);
+    }
+    // Data will move into the store, so we want to update dependencies before
+    // it is moved.
+    sink_->Init(data_set);
+    // After updating the sink let listeners know of changes.
+    if (change_notifications) {
+        NotifyChanges(*change_notifications);
+    }
+}
+
+void DataStoreUpdater::Upsert(std::string key,
+                              launchdarkly::server_side::data_source::
+                                  IDataSourceUpdateSink::FlagDescriptor flag) {
+    UpsertCommon(DataKind::kFlag, key, store_->GetFlag(key), flag);
```

Comment:
That was the plan, but I think it got lost in the layers.

Reply:
I may just make more things const and let it copy, otherwise it is really easy to use the key twice in one function and it to be invalid. Downside of the way C++ does moves.